internal/controlplane/handlers_githubwebhooks.go

```diff
@@ -186,6 +188,49 @@ func (s *Server) HandleGitHubWebHook() http.HandlerFunc {
 	}
 }
 
+func validatePayloadSignature(r *http.Request, wc *server.WebhookConfig) (payload []byte, err error) {
+	var bodyBytes []byte
+
+	// if there are any previous secrets we need to save the body
+	// because every call to ValidatePayload will consume the body
+	if len(wc.PreviousWebhookSecrets) > 0 {
+		bodyBytes, err = getAndDuplicateBody(r)
+		if err != nil {
+			return
+		}
+	}
+
+	payload, err = github.ValidatePayload(r, []byte(wc.WebhookSecret))
```

Comment:
There's a [`ValidatePayloadFromBody`](https://pkg.go.dev/github.com/google/go-github/v43/github#ValidatePayloadFromBody) which takes an `io.Reader` and the two header fields needed for the validation which might be simpler to use:

```suggestion
	bodyBytes, err := io.ReadAll(r.Body)
	if err != nil {
		return nil, err
	}
	body := io.NopCloser(bytes.NewReader(bodyBytes))
	signature := r.Header.Get(github.SHA256SignatureHeader)
	if signature == "" {
		signature = r.Header.Get(github.SHA1SignatureHeader)
	}

	contentType, _, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
	if err != nil {
		return nil, err
	}

	payload, err = github.ValidatePayload(contentType, body, signature, []byte(wc.WebhookSecret))
	
	body.Seek(0, io.SeekStart)
```

Reply:
yes, the reason I didn't use it was that I didn't want to copy reading the signature and the content type headers. I'll experiment with how the code feels before the next push along with the other changes.

Reply:
taken into account, it's not like the headers and the mime type are going to change, WDYT?